addons/membership/models/invoice.py

```diff
@@ -0,0 +1,78 @@
+# -*- coding: utf-8 -*-
+from datetime import date
+from odoo import api, models, fields, _
+import odoo.addons.decimal_precision as dp
+from odoo.exceptions import ValidationError
+
+
+class Invoice(models.Model):
+    _inherit = 'account.invoice'
+   
+    @api.multi
+    def action_cancel(self):
+        '''Create a 'date_cancel' on the membership_line object'''
+        res = super(Invoice, self).action_cancel()
+        today_date = date.today()
+        membership_lines = self.env['membership.membership_line'].search([('account_invoice_line_id', 'in', self.mapped('invoice_line_ids').ids)])
+        membership_lines.write({'date_cancel': today_date})
+        return res
+
+
+class AccountInvoiceLine(models.Model):
+    _inherit = 'account.invoice.line'
+   
+    @api.multi
+    def write(self, vals):
+        res = super(AccountInvoiceLine, self).write(vals)
+        MembershipLine = self.env['membership.membership_line']
+        for line in self.filtered(lambda l: l.invoice_id.type == 'out_invoice'):
+            membership_lines = MembershipLine.search([('account_invoice_line_id', '=', line.id)])
+            if line.product_id and line.product_id.membership and not membership_lines:
```

Comment:
No need to check this "line.product_id and"  in if condition simply : 
if line.product_id.membership and not membership_lines: